src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/LdapAccessPolicyDAO.scala

```diff
@@ -107,15 +112,28 @@ class LdapAccessPolicyDAO(protected val ldapConnectionPool: LDAPConnectionPool,
     newPolicy
   }
 
-  override def listAccessPolicies(resourceTypeName: ResourceTypeName, userId: WorkbenchUserId): Future[Set[ResourceIdAndPolicyName]] = Future {
-    val entry = ldapConnectionPool.getEntry(userDn(userId), Attr.memberOf)
-    val groupDns = getAttributes(entry, Attr.memberOf).getOrElse(Set.empty).toSet
+  override def listAccessPolicies(resourceTypeName: ResourceTypeName, userId: WorkbenchUserId): Future[Set[ResourceIdAndPolicyName]] = listAccessPoliciesIO(resourceTypeName, userId).unsafeToFuture()
 
-    val policyDnPattern = dnMatcher(List(Attr.policy, Attr.resourceId), resourceTypeDn(resourceTypeName))
+  private def unmarshalResourceIdAndPolicyName(entry: Entry, resourcesUserIsMember: Set[WorkbenchGroupName]): Either[String, ResourceIdAndPolicyName] = {
+    val authDomains = getAttributes(entry, Attr.authDomain).map(WorkbenchGroupName)
+    for{
+      resourceId <- getAttribute(entry, Attr.resourceId).toRight(s"${entry.getDN} attribute missing: ${Attr.resourceId}. ")
+      policyName <- getAttribute(entry, Attr.policy).toRight(s"${entry.getDN} attribute missing: ${Attr.policy}. ")
+    } yield ResourceIdAndPolicyName(ResourceId(resourceId), AccessPolicyName(policyName), authDomains, authDomains.filterNot(gn => resourcesUserIsMember.contains(gn)))
+  }
 
-    groupDns.collect {
-      case policyDnPattern(policyName, resourceId) => ResourceIdAndPolicyName(ResourceId(resourceId), AccessPolicyName(policyName))
-    }
+  protected[openam] def listAccessPoliciesIO(resourceTypeName: ResourceTypeName, userId: WorkbenchUserId): IO[Set[ResourceIdAndPolicyName]] = {
```

Comment:
my main concern with this is that it packs too much functionality into the DAO. The DAO should be about how to query, marshal and unmarshal the data. I think the composition of that logic should go into the ResourceService.